### PR TITLE
8115 parallel zfs mount (v2)

### PR DESCRIPTION
**Pull request by prakashsurya**
Reviewed by: Matthew Ahrens mahrens@delphix.com
Reviewed by: Pavel Zakharov pavel.zakharov@delphix.com
Reviewed by: Brad Lewis <brad.lewis@delphix.com>
Reviewed by: George Wilson <george.wilson@delphix.com>
Reviewed by: Paul Dagnelie <pcd@delphix.com>
Reviewed by: Prashanth Sreenivasa <pks@delphix.com>

Overview
========

In analyzing the time it takes for a Delphix Engine to come up following
a planned or unplanned reboot, we've determined that the SMF service
(filesystem/local) that's responsible for mounting all local filesystems
(except for /) is responsible for a significant percentage of the boot
time. The longer it takes for the Delphix Engine to come up, the longer
the Delphix Engine is unavailable during these outages. For example, on
a Delphix Engine with roughly 3000 filesystems, we have the following
breakdown of "filesystem/local" start time for a sample of 74 reboots:

    # NumSamples = 74; Min = 0.00; Max = 782.00
    # Mean = 186.972973; Variance = 17853.891161; SD = 133.618454; Median 156.000000
    # each * represents a count of 1
        0.0000 -    78.2000 [    10]: **********
       78.2000 -   156.4000 [    27]: ***************************
      156.4000 -   234.6000 [    17]: *****************
      234.6000 -   312.8000 [     8]: ********
      312.8000 -   391.0000 [     8]: ********
      391.0000 -   469.2000 [     1]: *
      469.2000 -   547.4000 [     1]: *
      547.4000 -   625.6000 [     1]: *
      625.6000 -   703.8000 [     0]:
      703.8000 -   782.0000 [     1]: *

On average, it takes over 3 minutes to mount local filesystems on that
system. A sampling of 56 reboots on another system which has 9000+
filesystems is below:

    # NumSamples = 56; Min = 0.00; Max = 1377.00
    # Mean = 175.250000; Variance = 54092.223214; SD = 232.577349; Median 118.000000
    # each * represents a count of 1
        0.0000 -   137.7000 [    37]: *************************************
      137.7000 -   275.4000 [    11]: ***********
      275.4000 -   413.1000 [     4]: ****
      413.1000 -   550.8000 [     1]: *
      550.8000 -   688.5000 [     1]: *
      688.5000 -   826.2000 [     0]:
      826.2000 -   963.9000 [     0]:
      963.9000 -  1101.6000 [     1]: *
     1101.6000 -  1239.3000 [     0]:
     1239.3000 -  1377.0000 [     1]: *

Mounting of filesystems in "filesystem/local" is done using `zfs mount -a`,
which mounts each filesystems serially. The bottleneck for each mount is
the I/O done to load metadata for each filesystem. As such, mounting
filesystems using a parallel algorithm should be a big win, and bring down
the runtime of "filesystem/local"'s start method.

Performance Testing: System Configuration
=========================================

To test and verify these changes impacted performance how we expected it
to, we used a VM with:

  - 8 vCPUs
  - zpool with 10 10k-SAS disks
  - filesystem hierarchy like so:

        1 pool     2 groups  100 containers  2 timeflows    5 leaf datasets
                               per group     per container  per timeflow
        ===================================================================
        test-pool-+-group-0-+-container-0-+---timeflow-0---+-ds-0
                  |         |             |                +-ds-1
                  |         |             |                +-ds-2
                  |         |             |                +-ds-3
                  |         |             |                +-ds-4
                  |         |             |
                  |         |             +---timeflow-1---+-ds-0
                  |         |                              +-ds-1
                  |         |                              +-ds-2
                  |         |                              +-ds-3
                  |         |                              +-ds-4
                  |         |
                  |         +-container-1-+---timeflow-0---+-ds-0
                  |         |             |                +-ds-1
                  |         |             |                +-ds-2
                  |         |             |                +-ds-3
                  |         |             |                +-ds-4
                  |         |             |
                  |         |             +---timeflow-1---+-ds-0
                  |         |                              +-ds-1
                  |         |                              +-ds-2
                  |         |                              +-ds-3
                  |         |                              +-ds-4
                  |         + ...
                  |         .
                  |         .
                  |
                  +-group-1 ...

This makes for a total of 2603 filesystems:

    pool + groups + containers + timeflows + leaves
    1    + 2      + 2*100      + 2(2*100)  + 5(2(2*100)) = 2603 filesystems

Additionally, a 1MB file was created in each leaf dataset.

Because this filesystem heirarchy is not very deep, this lends itself
well to the new parallel mounting algorithm implemented.

Performance Testing: Methodology and Results
============================================

The system described above was rebooted 10 times, and the duration of
the start method of "filesystem/local" was measured. Specifically, the
"zfs mount -va" comamnd that it calls was instrumented to break down the
phases of the mounting process into three buckets:

  1. gathering the list of filesystems to mount (aka "load")
  2. mounting all filesystems (aka "mount")
  3. left-over time spent doing anything else (aka "other")

The results of these measurements is below:

           | other (s) | load (s) | mount (s) |
       ----+-----------+----------+-----------+
    Before |    1.5    |    8.1   |    45.5   |
       ----+-------+------+-------+-----------+
     After |    1.7    |    7.9   |    2.1    |
       ----+-----------+----------+-----------+

In summary, for this configuration, the filesystem/local SMF services
goes from taking an average of 55.1 seconds (+/- 1.0s) to an average of
11.7 seconds (+/- 0.8s). The "other" and "load" times remain unchanged
(unsurprising given that this project hasn't touched any code in those
areas).

The big win comes in the "mount" phase, which reduces the time from
roughly 45 seconds to 2 seconds; a 95% decrease in latency.

Using the same zpool as above, "zpool import" performance was also
tested; the mounting done by "zpool import" now uses the same framework
as "zfs mount -a". Performance improvement for this case is unsurprisingly
on par with the "zfs mount -a" improvement documented above.

Upstream bugs: DLPX-46555, DLPX-49847, DLPX-49351, 38457

### CODE REVIEW

**Comment by prakashsurya** [comment]
@andy-js I'd appreciate your review on this, if you have the time.

I've rebased this onto the latest master changes, and the `libfakekernel` changes you made in #494. I ran into some build issues where some files would `#include <synch.h>` and this would conflict with symbols/macros from `libfakekernel`. As a workaround, I added an empty `synch.h` file in `libfakekernel`, but I'm not sure if this is the correct way to address the issue.

I still need to do some more testing, but it appears to work in the preliminary tests that I've done so far (boot, minimal zloop, mount specific tests from zfstest).

**Comment by rmustacc** [comment]
When trying to review this, I found the use of the fake kernel here really confusing. Why are we using it for a normal user land program? Doesn't it strike others as odd when you have a userland program use a `kmutex_t`, but then also referring to the userland bits like a normal USYNC_THREAD, which is definitely a userland mutex thing. It also makes it much less clear which mutex_enter/mutex_exit are we using. The real libc one or the renamed one in this kind of fake kernel world.

Unrelated to the above confusion, do you have an analysis somewhere about why we're marking specific file systems as requiring to be added to or removed from the mount in progress tables? It seems like this is probably an important part of this change. Can we get some comments as to why the set of file systems in question were noted as being safe to not be included in there.

**Comment by prakashsurya** [comment]
Thanks for giving this a look @rmustacc. 

> Doesn't it strike others as odd when you have a userland program use a kmutex_t, but then also referring to the userland bits like a normal USYNC_THREAD, which is definitely a userland mutex thing.

Thanks for catching this; this is wrong. Rather than using `USYNC_THREAD`, I think I should be using `MUTEX_DEFAULT`.

> It also makes it much less clear which mutex_enter/mutex_exit are we using.

AFAIK, there's no `mutex_enter`/`mutex_exit` provided by `libc`. Instead, there's `mutex_lock`/`mutex_unlock`. With that said, there is a `mutex_init` provided by `libc`, and then a `kmutex_init` provided by `libfakekernel`.

And then we use a macro to allow consumers use `mutex_init` when it's actually using `kmutex_init` (via `libfakekernel/common/sys/mutex.h`:
```
/*
 * We're simulating the kernel mutex API here, and the
 * user-level has a different signature, so rename.
 */
#define	mutex_init	kmutex_init
#define	mutex_destroy	kmutex_destroy
```

> Why are we using it for a normal user land program?

Since this change is using `libfakekernel`'s taskq implementation, I thought I had to use its mutex implementation as well, due to how it remaps functions like `mutex_init` to the `kmutex` variant. I'm not opposed to using the `libc` mutex implementation, so long as we can still use the other functionality from `libfakekernel` that is required.

Let me know if you think I should try using the `libc` implementation. I think I hit some build errors when doing that originally, but I can try again if you'd like me to.

> Unrelated to the above confusion, do you have an analysis somewhere about why we're marking specific file systems as requiring to be added to or removed from the mount in progress tables?

@sebroy can you comment on this?

**Comment by prakashsurya** [comment]
> Unrelated to the above confusion, do you have an analysis somewhere about why we're marking specific file systems as requiring to be added to or removed from the mount in progress tables? It seems like this is probably an important part of this change. Can we get some comments as to why the set of file systems in question were noted as being safe to not be included in there.

@rmustacc I'll open a separate PR for these changes. Looking at our internal JIRA, we have a separate issue for this with lots of details, so I think it makes sense to split them apart and land that independent of this feature. I'll work on doing that today.

**Comment by andy-js** [comment]
I think @rmustacc is exactly right.  We should not be building the zfs command (or libzfs) in fakekernel context.  This is a highly specialised environment that really only makes sense for things like libzpool which include kernel code.

I suggest you have a look at this earlier webrev of mine which introduces _TASKQUSER which can be used to access the taskq API outside if fakekernel context.  It was modelled after the _KMEMUSER define which does something similar:

http://cr.illumos.org/~webrev/andy_js/8115-1/

**Comment by prakashsurya** [comment]
Thanks. I'll look into that, and push a new revision when I have something working.

**Comment by prakashsurya** [comment]
@andy-js @rmustacc let me know if my latest change is what y'all were thinking. I haven't tested this yet, beyond verifying it builds.

**Comment by prakashsurya** [comment]
@andy-js @rmustacc is this good to RTI? it's passing the build, zloop, and zfstest.

**Comment by rmustacc** [comment]
@prakashsurya Thanks for splitting out the change regarding the file system part, that makes things clearer.

Honestly though, no, I still feel like this is the wrong approach in terms of leveraging the fake kernel library. But the fake kernel and libzpool are similar to me in so far as their goal is to emulate the kernel. This is a user land program which really shouldn't need any of these and while portability isn't my first concern, starts making things pretty odd for those porting this.

Honestly, if we take a step back at the original version, the only thing I was unhappy with was saying that we had a general purpose taskq user library because you weren't giving control to applications about how to handle memory allocation by using the umem fatal stuff. If we had scoped that to be private just for zfs or private in so far as just the fake kernel and the zfs user land commands using it, it probably would be a better approach. I'm sorry that I haven't been as active in the other parts. When I saw the fake kernel bits going on, I hadn't realized it was going on for use here.

I'm not sure if it makes sense. I see that there's now been Andy's changes and breaking it up further. But honestly, this still feels like a lot of complexity that we're introducing into the headers and builds and complications. I'll take a look at this current version and try to evaluate it. I'll try and hold my breath about the fake kernel bits.

File: usr/src/uts/common/sys/vfs.h [hunk 1]
```diff
@@ -416,6 +416,7 @@ enum {
 #define	VSW_XID		0x40	/* file system supports extended ids */
 #define	VSW_CANLOFI	0x80	/* file system supports lofi mounts */
 #define	VSW_ZMOUNT	0x100	/* file system always allowed in a zone */
+#define	VSW_MOUNTDEV	0x200	/* file system is mounted via device path */
```
Comment (rmustacc) [inline usr/src/uts/common/sys/vfs.h]: I thought we had said this was being pulled out into a separate change?

File: usr/src/cmd/zfs/zfs_main.c [hunk 2]
```diff
@@ -6163,33 +6199,44 @@ share_mount(int op, int argc, char **argv)
...
-			zfs_close(dslist[i]);
-		}
+		share_mount_state_t share_mount_state = { 0 };
+		share_mount_state.sm_op = op;
+		share_mount_state.sm_verbose = verbose;
+		share_mount_state.sm_flags = flags;
+		share_mount_state.sm_options = options;
+		share_mount_state.sm_proto = protocol;
+		share_mount_state.sm_total = cb.cb_used;
+		(void) mutex_init(&share_mount_state.sm_lock, USYNC_THREAD,
```
Comment (rmustacc) [inline usr/src/cmd/zfs/zfs_main.c]: Missing LOCK_ERRORCHECK

File: usr/src/lib/libzfs/common/libzfs_dataset.c [hunk 3]
```diff
@@ -785,6 +786,7 @@ libzfs_mnttab_cache_compare(const void *arg1, const void *arg2)
 void
 libzfs_mnttab_init(libzfs_handle_t *hdl)
 {
+	(void) mutex_init(&hdl->libzfs_mnttab_cache_lock, USYNC_THREAD, NULL);
```
Comment (rmustacc) [inline usr/src/lib/libzfs/common/libzfs_dataset.c]: Missing LOCK_ERRORCHECK

File: usr/src/lib/libzfs/common/libzfs_dataset.c [hunk 4]
```diff
@@ -854,16 +858,18 @@ libzfs_mnttab_find(libzfs_handle_t *hdl, const char *fsname,
 			return (ENOENT);
 	}
 
+	(void) mutex_lock(&hdl->libzfs_mnttab_cache_lock);
```
Comment (rmustacc) [inline usr/src/lib/libzfs/common/libzfs_dataset.c]: Use mutex_enter(), the version in user land. It will do all the error checking and abort appropriately. Otherwise, you need to check the return value and blow up if it's non-zero so we abort on deadlock.

File: usr/src/lib/libzfs/common/libzfs_dataset.c [hunk 5]
```diff
@@ -854,16 +858,18 @@ libzfs_mnttab_find(libzfs_handle_t *hdl, const char *fsname,
...
 
 	find.mtn_mt.mnt_special = (char *)fsname;
 	mtn = avl_find(&hdl->libzfs_mnttab_cache, &find, NULL);
 	if (mtn) {
 		*entry = mtn->mtn_mt;
-		return (0);
+		ret = 0;
 	}
-	return (ENOENT);
+	(void) mutex_unlock(&hdl->libzfs_mnttab_cache_lock);
```
Comment (rmustacc) [inline usr/src/lib/libzfs/common/libzfs_dataset.c]: Use mutex_exit(), the version in user land. It will do all the error checking and abort appropriately. Otherwise, you need to check the return value and blow up if it's non-zero so we abort on deadlock or unlocking a lock that we didn't own.

File: usr/src/lib/libzfs/common/libzfs_dataset.c [hunk 6]
```diff
@@ -872,14 +878,16 @@ libzfs_mnttab_add(libzfs_handle_t *hdl, const char *special,
...
 
-	if (avl_numnodes(&hdl->libzfs_mnttab_cache) == 0)
-		return;
-	mtn = zfs_alloc(hdl, sizeof (mnttab_node_t));
-	mtn->mtn_mt.mnt_special = zfs_strdup(hdl, special);
-	mtn->mtn_mt.mnt_mountp = zfs_strdup(hdl, mountp);
-	mtn->mtn_mt.mnt_fstype = zfs_strdup(hdl, MNTTYPE_ZFS);
-	mtn->mtn_mt.mnt_mntopts = zfs_strdup(hdl, mntopts);
-	avl_add(&hdl->libzfs_mnttab_cache, mtn);
+	(void) mutex_lock(&hdl->libzfs_mnttab_cache_lock);
```
Comment (rmustacc) [inline usr/src/lib/libzfs/common/libzfs_dataset.c]: Use mutex_enter(), the version in user land. It will do all the error checking and abort appropriately. Otherwise, you need to check the return value and blow up if it's non-zero so we abort on deadlock.

File: usr/src/lib/libzfs/common/libzfs_dataset.c [hunk 7]
```diff
@@ -872,14 +878,16 @@ libzfs_mnttab_add(libzfs_handle_t *hdl, const char *special,
...
+	if (avl_numnodes(&hdl->libzfs_mnttab_cache) != 0) {
+		mtn = zfs_alloc(hdl, sizeof (mnttab_node_t));
+		mtn->mtn_mt.mnt_special = zfs_strdup(hdl, special);
+		mtn->mtn_mt.mnt_mountp = zfs_strdup(hdl, mountp);
+		mtn->mtn_mt.mnt_fstype = zfs_strdup(hdl, MNTTYPE_ZFS);
+		mtn->mtn_mt.mnt_mntopts = zfs_strdup(hdl, mntopts);
+		avl_add(&hdl->libzfs_mnttab_cache, mtn);
+	}
+	(void) mutex_unlock(&hdl->libzfs_mnttab_cache_lock);
```
Comment (rmustacc) [inline usr/src/lib/libzfs/common/libzfs_dataset.c]: Use mutex_exit(), the version in user land. It will do all the error checking and abort appropriately. Otherwise, you need to check the return value and blow up if it's non-zero so we abort on deadlock or unlocking a lock that we didn't own.

File: usr/src/lib/libzfs/common/libzfs_dataset.c [hunk 8]
```diff
@@ -888,6 +896,7 @@ libzfs_mnttab_remove(libzfs_handle_t *hdl, const char *fsname)
 	mnttab_node_t find;
 	mnttab_node_t *ret;
 
+	(void) mutex_lock(&hdl->libzfs_mnttab_cache_lock);
```
Comment (rmustacc) [inline usr/src/lib/libzfs/common/libzfs_dataset.c]: Same mutex comments.

File: usr/src/lib/libzfs/common/libzfs_dataset.c [hunk 9]
```diff
@@ -898,6 +907,7 @@ libzfs_mnttab_remove(libzfs_handle_t *hdl, const char *fsname)
 		free(ret->mtn_mt.mnt_mntopts);
 		free(ret);
 	}
+	(void) mutex_unlock(&hdl->libzfs_mnttab_cache_lock);
```
Comment (rmustacc) [inline usr/src/lib/libzfs/common/libzfs_dataset.c]: Same mutex comments.

File: usr/src/lib/libzfs/common/libzfs_mount.c [hunk 10]
```diff
@@ -88,6 +89,9 @@
 #include <sys/systeminfo.h>
 #define	MAXISALEN	257	/* based on sysinfo(2) man page */
 
+static int mount_tq_nthr = 512;	/* taskq threads for multi-threaded mounting */
```
Comment (rmustacc) [inline usr/src/lib/libzfs/common/libzfs_mount.c]: How did we determine this number? How does this change if you're not in the global zone trying to mount a lot of things and you have CPU caps that limit the amount of parallelism?

File: usr/src/lib/libzfs/common/libzfs_mount.c [hunk 11]
```diff
@@ -1121,104 +1132,349 @@ mount_cb(zfs_handle_t *zhp, void *data)
...
+
+	mnt_param->mnt_hdl = hdl;
+	mnt_param->mnt_tq = tq;
+	mnt_param->mnt_zhps = handles;
+	mnt_param->mnt_num_handles = num_handles;
+	mnt_param->mnt_idx = idx;
+	mnt_param->mnt_func = func;
+	mnt_param->mnt_data = data;
+
+	(void) taskq_dispatch(tq, zfs_mount_task, (void*)mnt_param, TQ_SLEEP);
```
Comment (rmustacc) [inline usr/src/lib/libzfs/common/libzfs_mount.c]: In this case if we run low on memory this means we'll abort, right? Is that what we want?

File: usr/src/uts/common/fs/hsfs/hsfs_vfsops.c [hunk 12]
```diff
@@ -150,7 +151,7 @@ static vfsdef_t vfw = {
 	"hsfs",
 	hsfsinit,
 	/* We don't suppport remounting */
-	VSW_HASPROTO|VSW_STATS|VSW_CANLOFI,
+	VSW_HASPROTO|VSW_STATS|VSW_CANLOFI|VSW_MOUNTDEV,
```
Comment (rmustacc) [inline usr/src/uts/common/fs/hsfs/hsfs_vfsops.c]: Here and all the others, aren't these supposed to be pulled out?

**Comment by prakashsurya** [comment]
@rmustacc Thanks for the reply. Let me know after you give this version a look. If you think `libzfs` being a consumer of `libfakekernel`'s taskq implementation, then let's discuss the best way to break that out so we can do this properly.

File: usr/src/uts/common/sys/vfs.h (hunk 1 above)
Comment (prakashsurya) [reply usr/src/uts/common/sys/vfs.h]: Yes, this is in the PR here: https://github.com/openzfs/openzfs/pull/538

Since that PR is a pre-requisite of this ZFS change, I have both changes in this diff. If you only look at the second commit of the PR, [here](https://github.com/openzfs/openzfs/pull/536/commits/b3f4bdf804d187b4d17d6721d3b47c2e53685e07), that'll contain only the ZFS changes.

Sorry, Github's PR system doesn't have a way to chain dependent PRs, AFAIK.

File: usr/src/lib/libzfs/common/libzfs_mount.c (hunk 10 above)
Comment (prakashsurya) [reply usr/src/lib/libzfs/common/libzfs_mount.c]: @sebroy can you comment here? I don't see anything in the original review about this.

File: usr/src/lib/libzfs/common/libzfs_mount.c (hunk 11 above)
Comment (prakashsurya) [reply usr/src/lib/libzfs/common/libzfs_mount.c]: @ahrens how do we handle allocation failures elsewhere in the zfs userland code?

File: usr/src/lib/libzfs/common/libzfs_mount.c (hunk 10 above)
Comment (prakashsurya) [reply usr/src/lib/libzfs/common/libzfs_mount.c]: I talked to Seb about this today, and it's like I was assuming.

The number 512 here isn't necessarily special (we could change it), we just want a number large enough such that we can create enough threads to saturate the storage IOPs. Since each thread performing a mount will generally be IO bound, we want to have enough threads mounting concurrently, so that we more effectively utilize the storage IOPs.

As far as CPU caps and zone considerations, I _think_ 512 should be fine. The performance consideration here is to make sure we issue enough IO, so we still want to have multiple threads issues mounts concurrently, even if we can only have a single thread running at a time. i.e. we want to be able to issue another mount, when the previous thread/mount blocks on IO.

With that said, I'm not familiar with the CPU caps available, so perhaps I'm overlooking something. Let me know if you'd like some more information, and/or if you still have concerns here.

File: usr/src/lib/libzfs/common/libzfs_mount.c (hunk 11 above)
Comment (ahrens) [reply usr/src/lib/libzfs/common/libzfs_mount.c]: From looking at the code, if libzfs_printerr is set then libzfs will call exit() on memory allocation failure.  This is not set by default, but it is set by the zfs, zpool, zinject, libshare, dumpadm, and swap commands.  So for the common ways of hitting this code ("zfs mount", "zfs share", "zpool import"), libzfs_printerr will be set, indicating that the desired behavior is to exit.  The only caller of this code that might not have libzfs_printerr set is syseventd (it does this when a pool goes from being FAULTED (or worse) to DEGRADED or HEALTHY).

In practice, it looks like lots of (but not all) callers are assuming that we will die if allocation fails (see callers of no_memory(), like zfs_alloc() and the comments above them - the assume that libzfs_printerr is always set so we always exit on failure).

So I'd say this is in keeping with the current behavior of handling allocation failures poorly :-(

**Comment by prakashsurya** [comment]
@rmustacc any more questions/comments on this?

File: usr/src/lib/libzfs/common/sys/zfs_context.h [hunk 13]
```diff
@@ -1,37 +0,0 @@
...
-#include <sys/taskq_impl.h>
-#include <sys/debug.h>
-#include <sys/list.h>
-#include <sys/zfs_debug.h>
-
-#ifdef	__cplusplus
-}
-#endif
-
-#endif	/* _SYS_ZFS_CONTEXT_H */
```
Comment (andy-js) [inline usr/src/lib/libzfs/common/sys/zfs_context.h]: I'm glad to see this file go.  Too bad we can't also drop the one in libzpool.

File: usr/src/lib/libzfs/common/libzfs_mount.c (hunk 10 above)
Comment (andy-js) [reply usr/src/lib/libzfs/common/libzfs_mount.c]: I suppose this could be made a tuneable through /etc/default/zfs?

File: usr/src/lib/libzfs/common/libzfs_mount.c (hunk 11 above)
Comment (rmustacc) [reply usr/src/lib/libzfs/common/libzfs_mount.c]: OK, thanks for clarifying that. That's unfortunate, but life.

File: usr/src/lib/libzfs/common/libzfs_mount.c (hunk 10 above)
Comment (rmustacc) [reply usr/src/lib/libzfs/common/libzfs_mount.c]: Well, it's pretty easy to go tune down the max lwps you have in a zone or similar. Though hopefully in those situations you won't have a lot of datasets. It's a bit unfortunate that the model doesn't fall back to the known single threaded bit on resource exhaustion. Hopefully we won't have people hitting this substantially.

**Comment by rmustacc** [comment]
@prakashsurya I do think libzfs in general being a consumer of a fake kernel library is the wrong thing. I mean, wouldn't you say the same thing if it was linked against libzpool? It feels like we're going to all these contortions to emulate a kernel environment which we don't do for other code and things that are shared between user land and the kernel. Even things like ctf or avl trees have code in common that builds in either environment and glue in the user land and the kernel.

Again, I realize I'm probably to blame for how we got here since you all first proposed a generic taskq API that looked like the kernels, but basically made it impossible for the non-libzfs panic on out of memory case. It would probably have been better to have that and have the fake kernel bits rely on that same thing, then the other way around. But everyone's already done the work in the other direction, which is unfortunate, but reality. I'm not sure I can sign off as being a reviewer on this.

I can understand that folks don't want to change, but then I would request that you make it very clear to the RTI advocates that there are actual concerns with this from an architectural perspective. I would bet that changes to the fake kernel code and header games are going to impact this in the future.

**Comment by prakashsurya** [comment]
@rmustacc OK. So, if I'm understanding you correctly, you'd like us to take the taskq implementation from libfakekernel, move it into a new userspace library (e.g. libutaskq), and then make libfakekernel and libzfs consume this new library?

**Comment by rmustacc** [comment]
On 2/16/18 9:17 , Prakash Surya wrote:
> @rmustacc OK. So, if I'm understanding you correctly, you'd like us to take the taskq implementation from libfakekernel, move it into a new userspace library (e.g. libutaskq), and then make libfakekernel and libzfs consume this new library?
> 

I think that both matches what you would have originally done and it'd
probably make more sense from a design perspective. But I also don't
want to be the one forcing everyone to do more busy work if I'm the only
one who thinks that.

**Comment by andy-js** [comment]
@prakashsurya You may count me as a reviewer if you wish to proceed down the current path, though I am having some seconds thoughts based on Robert's comments.  We could break this up in a subsequent change I guess.  I'd be okay with that.

**Comment by rmustacc** [comment]
On 2/16/18 9:17 , Prakash Surya wrote:
> @rmustacc OK. So, if I'm understanding you correctly, you'd like us to take the taskq implementation from libfakekernel, move it into a new userspace library (e.g. libutaskq), and then make libfakekernel and libzfs consume this new library?
> 

I guess I should also add that something like libfakekernel is probably
going to be a PITA for any other ZFS consumer, but don't kow that with
certainty.

**Comment by prakashsurya** [comment]
I will investigate the work needed to break the taskq implementation out of libfakekernel, and into a new libutaskq library. I'll report back when I have something reasonable to show. Thanks for all the input! I appreciate it.

**Comment by prakashsurya** [comment]
In my latest commit, I've pulled in the taskq implementation that we've been using in delphix-os (via libcmdutils), directly into libzfs (with slight modifications).

It might be possible to pull this out into a completely separate (and new) library, but I'm not sure that's actually useful, since there's some outstanding concerns w.r.t. out-of-memory handling, and based on @gwr's comment.

For now, I've left symbols with the `utaskq` prefix, but if we all agree that it's OK to add this implementation directly to libzfs, and *not* use libfakekernel's taskq implementation, I think it might be worthwhile to change this prefix to something more libzfs specific. E.g. rename `utaskq.c` to `libzfs_taskq.c`, rename `utaskq_create` to `libzfs_taskq_create`, etc.

Folks, please let me know what y'all think.

**Comment by prakashsurya** [comment]
Thanks @gwr.

I was planning to have libzpool continue to use the taskq implementation in libfakekernel, and only have libzfs use this new implementation (since we don't want libzfs to use the libfakekernel one). Does that sound OK to you?

I appreciate the input. I'll try and get a cleaned up and tested version of this out for review in the coming weeks.

**Comment by andy-js** [comment]
I can live with this.

Rather than having 3 userspace taskq implementations (the one in libzpool, libfakekernel and libutaskq) we now have 2 (one in libfakekernel and one in libzfs).  That's an improvement.

**Comment by prakashsurya** [comment]
This is superseded by #596
